devito/builtins.py

```diff
@@ -53,6 +100,212 @@ def smooth(f, g, axis=None):
+
+    mapper = {}
+    for d in f_c.dimensions:
+        lhs = []
+        rhs = []
+        options = []
+
+        lhs.append(f_o)
+        rhs.append(dv.generic_derivative(f_c, d, 2*lw, 1))
+        coeffs = dv.Coefficient(1, f_c, d, weights)
+        options.append({'coefficients': dv.Substitutions(coeffs),
+                        'subdomain': grid.subdomains['objective_domain']})
+        lhs.append(f_c)
+        rhs.append(f_o)
+        options.append({'subdomain': grid.subdomains['objective_domain']})
+
+        mapper[d] = {'lhs': lhs, 'rhs': rhs, 'options': options}
+
+    initialize_function(f_c, f.data[:], lw,
+                        mapper=mapper,
+                        mode='reflect', name='smooth')
+
+    fset(f, f_c)
+
+    return f
+
+
+def initialize_function(function, data, nbl, mapper=dict(), mode='constant',
+                        name='padfunc'):
+    """
+    Initialize a `Function` with the given ``data``. ``data``
+    does *not* include the ``nbl`` outer/boundary layers; these are added via padding
+    by this function.
+
+    Parameters
+    ----------
+    function : Function
+        The initialised object.
+    data : ndarray of Function
+        The data used for initialisation.
+    nbl : int
+        Number of outer layers (such as PML layers for boundary damping).
+    mapper : dict, optional
+        Dictionary containing, for each dimension of function, a sub-dictionary
+        containing the following keys:
+        1) 'lhs': List of additional expressions to be added to the LHS expressions list.
+        2) 'rhs': List of additional expressions to be added to the RHS expressions list.
+        3) 'options': Options pertaining to the additional equations that will be
+        constructed.
+    mode : str, optional
+        The function initialisation mode. 'constant' and 'reflect' are
+        accepted.
+    name : str, optional
+        The name assigned to the operator.
+
+    Examples
+    --------
+    In the following example the `inner` region of a function is set to one plus the
+    value on the boundary.
+
+    >>> import numpy as np
+    >>> from devito import Grid, SubDomain, Function, initialize_function
+
+    Create a subdomain representing the `inner` region:
+
+    >>> class Inner(SubDomain):
+    ...     name = 'inner'
+    ...     def define(self, dimensions):
+    ...        return {d: ('middle', 1, 1) for d in dimensions}
+    >>> inner = Inner()
+
+    Now create the computational domain including the `inner` subdomain:
+
+    >>> grid = Grid(shape=(6, 6), subdomains=inner)
+    >>> x, y = grid.dimensions
+
+    Create the function we wish to set along with the data we wish to use to set it:
+
+    >>> f = Function(name='f', grid=grid, dtype=np.int32)
+    >>> data = np.full((4, 4), 2, dtype=np.int32)
+
+    Create the additional expressions and options required to set the value of
+    the inner region to one greater than the boundary value. Note that the equation
```

Comment:
> Note that the equation is specified on the second grid dimension so that one is added to the inner values after all boundary values have been set.

This behaviour is not very clear to me. What would happen if you had used `x` in place of `y` ?

Reply:
You'd pad in the `y` direction with values of 1 greater than those padded in the `x` direction.